README.md
# Suspenders (私家版)

[thoughtbot](http://thoughtbot.com/community)が公開している
[Suspenders](https://github.com/thoughtbot/suspenders)をForkして自分用にカスタマイズしています。


## インストール方法

```
$ git clone git@github.com:bojovs/suspenders.git
$ gem install suspenders-1.1.5.gem
```


## 使い方

```
$ suspenders projectname
```

上記コマンドを実行すると `projectname` という名前の Rails 3.2 なアプリケーションが生成されます。


## やっていること

* 標準でインストールしたいgemsをGemfileに追加
  * インストール内容は[template/Gemfile_additions](/bojovs/suspenders/blob/master/templates/Gemfile_additions)を見てください。
* Override recipient emails in staging environment.
* Rails' flashes set up and in application layout.
* A few nice time formats set up for localization.
* [Heroku-recommended asset pipeline
  settings](https://devcenter.heroku.com/articles/rails3x-asset-pipeline-cedar/).


Heroku
------

You can optionally create Heroku staging and production apps:

    suspenders app --heroku true

This has the same effect as running:

    heroku create app-staging --remote staging
    heroku create app-production --remote production

Github
------

You can optionally create a Github repository:

    suspenders app --github organization/project

This has the same effect as running:

    hub create organization/project

Clearance
---------

You can optionally not include Clearance:

    suspenders app --clearance false

Capybara Webkit
---------------

You can optionally not include Capybara Webkit (which depends on QT being
installed on your machine):

    suspenders app --webkit false

Dependencies
------------

Some gems included in Suspenders have native extensions. You should have GCC installed on your
machine before generating an app with Suspenders.

Use [OS X GCC Installer](/kennethreitz/osx-gcc-installer/) for Snow Leopard
(OS X 10.6).

Use [Command Line Tools for XCode](https://developer.apple.com/downloads/index.action)
for Lion (OS X 10.7) or Mountain Lion (OS X 10.8).

We use [Capybara Webkit](/thoughtbot/capybara-webkit) for full-stack Javascript
integration testing. It requires QT. Instructions for installing QT are
[here](/thoughtbot/capybara-webkit/wiki/Installing-Qt-and-compiling-capybara-webkit).

PostgreSQL needs to be installed and running for the `db:create` rake task.

Issues
------

If you have problems, please create a [Github Issue](/thoughtbot/suspenders/issues).

Contributing
------------

Please see CONTRIBUTING.md for details.

Credits
-------

![thoughtbot](http://thoughtbot.com/images/tm/logo.png)

Suspenders is maintained and funded by [thoughtbot, inc](http://thoughtbot.com/community)

The names and logos for thoughtbot are trademarks of thoughtbot, inc.

License
-------

Suspenders is Copyright © 2008-2012 thoughtbot. It is free software, and may be
redistributed under the terms specified in the LICENSE file.


NEWS.md
1.1.5.1 (January 6, 2013)

* Forkしました。
* Cucumberのテストコードを削除しちゃいました。

1.1.5 (October 22, 2012)

* Ignore `.env`.
* Link to thoughtbot/guides in generated README.
* Remove Cucumber in favor of RSpec + Capybara.
* Deliver emails in staging environment to an overriden email or set of emails.
* Encode database as UTF8.
* Bundle with binstubs using 37signals' directory convention.
* Configure time formats using localization.
* Add Ruby version to Gemfile.
* Add fast-failing spec that tests validity of factories.
* Use SimpleCov for C0 coverage.
* Configure RSpec with `--profile` flag to find slow-running specs.

1.1.4 (September 4, 2012)

* Always store UTC in the DB.
* Use Rails 3.2.8.

1.1.3 (August 7, 2012)

* Fix broken Gemfile additions where capybara stole cucumber's `require: false`

1.1.2 (August 6, 2012)

* Fix broken rake.
* Use Heroku-compliant asset pipeline settings.

1.1.1 (August 3, 2012)

* Fix broken newline interpolation

1.1.0 (August 3, 2012)

* Add --github option.
* Add --webkit option.
* Remove cruft when generating controllers.
* Add Spork and Guard.

1.0.1 (August 2, 2012)

* Fix broken install on Ruby 1.8.
* Remove db/schema.rb from .gitignore.
* Remove Factory Girl step definitions.

1.0.0 (June 29, 2012)

* Ignore `bin/`, `.rake_tasks`, `.bundle/`, and `.gem/`.
* Create a root route only with `--clearance`.
* Do not autocommit after generate.
* Style static error pages.
* Cucumber requires everything under `features/`, regardless of pwd.
* Added gems:
  * `foreman`
  * `therubyracer`
* Removed gems:
  * `copycopter_client`
  * `heroku`
  * `ruby-debug`
  * `sass`
  * `sprockets-redirect`
  * `email_spec`


templates/email_validator.rb
# Example use:
#   validates :email, email: true

class EmailValidator < ActiveModel::EachValidator
  def validate_each(record, attribute, value)
    @record, @attribute, @value = record, attribute, value
    @email = Mail::Address.new(@value)
    @tree  = @email.__send__(:tree)
    add_error unless valid?
  rescue Mail::Field::ParseError
    add_error
  end

  private

  def valid?
    !!(domain_and_address_present? && domain_has_more_than_one_atom?)
  end

  def domain_and_address_present?
    @email.domain && @email.address == @value
  end

  def domain_has_more_than_one_atom?
    @tree.domain.dot_atom_text.elements.length > 1
  end

  def add_error
    if message = options[:message]
      @record.errors[@attribute] << message
    else
      @record.errors.add(@attribute, :invalid)
    end
  end
end


lib/suspenders/generators/app_generator.rb
require 'rails/generators'
require 'rails/generators/rails/app/app_generator'

module Suspenders
  class AppGenerator < Rails::Generators::AppGenerator
    class_option :clearance, :type => :boolean, :aliases => '-C', :default => true,
      :desc => 'Add the Clearance Rails authentication library'

    class_option :database, :type => :string, :aliases => '-d', :default => 'postgresql',
      :desc => "Preconfigure for selected database (options: #{DATABASES.join('/')})"

    class_option :heroku, :type => :boolean, :aliases => '-H', :default => false,
      :desc => 'Create staging and production Heroku apps'

    class_option :github, :type => :string, :aliases => '-G', :default => nil,
      :desc => 'Create Github repository and add remote origin pointed to repo'

    class_option :skip_test_unit, :type => :boolean, :aliases => '-T', :default => true,
      :desc => 'Skip Test::Unit files'

    class_option :webkit, :type => :boolean, :aliases => '-W', :default => true,
      :desc => 'Add the Capybara Webkit Javascript integration testing library'

    def finish_template
      invoke :suspenders_customization
      super
    end

    def suspenders_customization
      invoke :remove_files_we_dont_need
      invoke :setup_development_environment
      invoke :setup_staging_environment
      invoke :create_suspenders_views
      invoke :create_common_javascripts
      invoke :add_jquery_ui
      invoke :customize_gemfile
      invoke :setup_database
      invoke :configure_app
      invoke :setup_stylesheets
      invoke :copy_miscellaneous_files
      invoke :customize_error_pages
      invoke :remove_routes_comment_lines
      invoke :setup_root_route
      invoke :setup_git
      invoke :create_heroku_apps
      invoke :create_github_repo
      invoke :outro
    end

    def remove_files_we_dont_need
      build :remove_public_index
      build :remove_rails_logo_image
    end

    def setup_development_environment
      say 'Setting up the development environment'
      build :raise_delivery_errors
      build :enable_factory_girl_syntax
    end

    def setup_staging_environment
      say 'Setting up the staging environment'
      build :setup_staging_environment
      build :initialize_on_precompile
    end

    def create_suspenders_views
      say 'Creating suspenders views'
      build :create_partials_directory
      build :create_flashes_partial
      build :create_javascripts_partial
      build :create_application_layout
    end

    def create_common_javascripts
      say 'Pulling in some common javascripts'
      build :create_common_javascripts
    end

    def add_jquery_ui
      say 'Add jQuery ui to the standard application.js'
      build :add_jquery_ui
    end

    def customize_gemfile
      build :add_custom_gems

      if options[:clearance]
        build :add_clearance_gem
      end

      if options[:webkit]
        build :add_capybara_webkit_gem
      end

      bundle_command 'install'
    end

    def setup_database
      say 'Setting up database'

      if 'postgresql' == options[:database]
        build :use_postgres_config_template
      end

      build :create_database
    end

    def configure_app
      say 'Configuring app'
      build :configure_rspec
      build :configure_action_mailer
      build :generate_rspec
      build :configure_time_zone

      if options[:webkit]
        build :configure_capybara_webkit
      end

      build :setup_guard_spork
      build :add_email_validator
      build :setup_default_rake_task
      build :setup_clearance
    end

    def setup_clearance
      if options[:clearance]
        build :generate_clearance
        build :include_clearance_matchers

        if using_active_record?
          build :set_attr_accessibles_on_user
        end
      end
    end

    def setup_stylesheets
      say 'Set up stylesheets'
      build :setup_stylesheets
    end

    def setup_git
      say 'Initializing git'
      invoke :setup_gitignore
      invoke :init_git
    end

    def create_heroku_apps
      if options[:heroku]
        say 'Creating Heroku apps'
        build :create_heroku_apps
      end
    end

    def create_github_repo
      if options[:github]
        say 'Creating Github repo'
        build :create_github_repo, options[:github]
      end
    end

    def setup_gitignore
      build :gitignore_files
    end

    def init_git
      build :init_git
    end

    def copy_miscellaneous_files
      say 'Copying miscellaneous support files'
      build :copy_miscellaneous_files
    end

    def customize_error_pages
      say 'Customizing the 500/404/422 pages'
      build :customize_error_pages
    end

    def remove_routes_comment_lines
      build :remove_routes_comment_lines
    end

    def setup_root_route
      if options[:clearance]
        say 'Setting up a root route'
        build :setup_root_route
      end
    end

    def outro
      say 'Congratulations! You just pulled our suspenders.'
      say "Remember to run 'rails generate airbrake' with your API key."
    end

    def run_bundle
      # Let's not: We'll bundle manually at the right spot
    end

    protected

    def get_builder_class
      Suspenders::AppBuilder
    end

    def using_active_record?
      !options[:skip_active_record]
    end
  end
end


lib/suspenders/app_builder.rb
module Suspenders
  class AppBuilder < Rails::AppBuilder
    include Suspenders::Actions

    def readme
      template 'README.md.erb', 'README.md'
    end

    def remove_public_index
      remove_file 'public/index.html'
    end

    def remove_rails_logo_image
      remove_file 'app/assets/images/rails.png'
    end

    def raise_delivery_errors
      replace_in_file 'config/environments/development.rb',
        'raise_delivery_errors = false', 'raise_delivery_errors = true'
    end

    def provide_setup_script
      copy_file 'bin_setup', 'bin/setup'
      run 'chmod a+x bin/setup'
    end

    def enable_factory_girl_syntax
      copy_file 'factory_girl_syntax_rspec.rb', 'spec/support/factory_girl.rb'
    end

    def test_factories_first
      copy_file 'factories_spec.rb', 'spec/models/factories_spec.rb'
      append_file 'Rakefile', factories_spec_rake_task
    end

    def setup_staging_environment
      run 'cp config/environments/production.rb config/environments/staging.rb'
      inject_into_file 'config/environments/staging.rb',
        "\n  config.action_mailer.delivery_method = :override_recipient_smtp, to: 'staging@example.com'",
        :after => 'config.action_mailer.raise_delivery_errors = false'
    end

    def initialize_on_precompile
      inject_into_file 'config/application.rb',
        "\n    config.assets.initialize_on_precompile = false",
        :after => 'config.assets.enabled = true'
    end

    def create_partials_directory
      empty_directory 'app/views/application'
    end

    def create_shared_flashes
      copy_file '_flashes.html.erb', 'app/views/application/_flashes.html.erb'
    end

    def create_shared_javascripts
      copy_file '_javascript.html.erb', 'app/views/application/_javascript.html.erb'
    end

    def create_application_layout
      template 'suspenders_layout.html.erb.erb',
        'app/views/layouts/application.html.erb',
        :force => true
    end

    def create_common_javascripts
      directory 'javascripts', 'app/assets/javascripts'
    end

    def add_jquery_ui
      inject_into_file 'app/assets/javascripts/application.js',
        "//= require jquery-ui\n", :before => '//= require_tree .'
    end

    def use_postgres_config_template
      template 'postgresql_database.yml.erb', 'config/database.yml',
        :force => true
    end

    def create_database
      bundle_command 'exec rake db:create'
    end

    def set_ruby_to_version_being_used
      inject_into_file 'Gemfile', "\n\nruby '#{RUBY_VERSION}'",
        :after => /source 'https:\/\/rubygems.org'/
    end

    def add_custom_gems
      additions_path = find_in_source_paths('Gemfile_additions')
      new_gems = File.open(additions_path).read
      inject_into_file 'Gemfile', "\n#{new_gems}",
        :after => /gem 'jquery-rails'/
    end

    def add_clearance_gem
      inject_into_file 'Gemfile', "\ngem 'clearance'",
        :after => /gem 'jquery-rails'/
    end

    def add_capybara_webkit_gem
      inject_into_file 'Gemfile', "\n  gem 'capybara-webkit'",
        :after => /gem 'capybara'/
    end

    def configure_rspec
      remove_file '.rspec'
      copy_file 'rspec', '.rspec'
      prepend_file 'spec/spec_helper.rb', simplecov_init
      replace_in_file 'spec/spec_helper.rb',
        '# config.mock_with :mocha',
        'config.mock_with :mocha'

      rspec_expect_syntax = <<-RUBY

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
      RUBY

      generators_config = <<-RUBY
    config.generators do |generate|
      generate.test_framework :rspec
      generate.helper false
      generate.stylesheets false
      generate.javascript_engine false
      generate.view_specs false
    end
      RUBY

      inject_into_file 'spec/spec_helper.rb', rspec_expect_syntax,
        :after => 'RSpec.configure do |config|'
      inject_into_class 'config/application.rb', 'Application', generators_config
    end

    def configure_time_zone
      time_zone_config = <<-RUBY
          config.active_record.default_timezone = :utc
      RUBY
      inject_into_class "config/application.rb", "Application", time_zone_config
    end

    def configure_time_formats
      remove_file 'config/locales/en.yml'
      copy_file 'config_locales_en.yml', 'config/locales/en.yml'
    end

    def configure_action_mailer
      action_mailer_host 'development', "#{app_name}.local"
      action_mailer_host 'test', 'www.example.com'
      action_mailer_host 'staging', "staging.#{app_name}.com"
      action_mailer_host 'production', "#{app_name}.com"
    end

    def generate_rspec
      generate 'rspec:install'
    end

    def configure_capybara_webkit
      append_file 'spec/spec_helper.rb' do
        "\nCapybara.javascript_driver = :webkit"
      end
    end

    def setup_guard_spork
      copy_file 'Guardfile', 'Guardfile'
    end

    def generate_clearance
      generate 'clearance:install'
    end

    def setup_foreman
      copy_file 'sample.env', 'sample.env'
      copy_file 'Procfile', 'Procfile'
    end

    def setup_stylesheets
      copy_file 'app/assets/stylesheets/application.css',
        'app/assets/stylesheets/application.css.scss'
      remove_file 'app/assets/stylesheets/application.css'
      concat_file 'import_scss_styles', 'app/assets/stylesheets/application.css.scss'
      create_file 'app/assets/stylesheets/_screen.scss'
    end

    def gitignore_files
      concat_file 'suspenders_gitignore', '.gitignore'
      [
        'app/models',
        'app/assets/images',
        'app/views/pages',
        'db/migrate',
        'log',
        'spec/support',
        'spec/lib',
        'spec/models',
        'spec/views',
        'spec/controllers',
        'spec/helpers',
        'spec/support/matchers',
        'spec/support/mixins',
        'spec/support/shared_examples'
      ].each do |dir|
        empty_directory_with_gitkeep dir
      end
    end

    def init_git
      run 'git init'
    end

    def create_heroku_apps
      path_addition = override_path_for_tests
      run "#{path_addition} heroku create #{app_name}-production --remote=production"
      run "#{path_addition} heroku create #{app_name}-staging --remote=staging"
    end

    def create_github_repo(repo_name)
      path_addition = override_path_for_tests
      run "#{path_addition} hub create #{repo_name}"
    end

    def copy_libraries
      copy_file 'override_recipient_smtp.rb', 'lib/override_recipient_smtp.rb'
    end

    def copy_miscellaneous_files
      copy_file 'errors.rb', 'config/initializers/errors.rb'
    end

    def customize_error_pages
      meta_tags =<<-EOS
  <meta charset='utf-8' />
  <meta name='ROBOTS' content='NOODP' />
      EOS
      style_tags =<<-EOS
<link href='/assets/application.css' media='all' rel='stylesheet' type='text/css' />
      EOS
      %w(500 404 422).each do |page|
        inject_into_file "public/#{page}.html", meta_tags, :after => "<head>\n"
        replace_in_file "public/#{page}.html", /<style.+>.+<\/style>/mi, style_tags.strip
        replace_in_file "public/#{page}.html", /<!--.+-->\n/, ''
      end
    end

    def setup_root_route
      route "root :to => 'Clearance::Sessions#new'"
    end

    def remove_routes_comment_lines
      replace_in_file 'config/routes.rb',
        /Application\.routes\.draw do.*end/m,
        "Application.routes.draw do\nend"
    end

    def set_attr_accessibles_on_user
      inject_into_file 'app/models/user.rb',
        "  attr_accessible :email, :password\n",
        :after => /include Clearance::User\n/
    end

    def add_email_validator
      copy_file 'email_validator.rb', 'app/validators/email_validator.rb'
    end

    def include_clearance_matchers
      create_file 'spec/support/clearance.rb', "require 'clearance/testing'"
    end

    def setup_default_rake_task
      append_file 'Rakefile' do
        "task(:default).clear\ntask :default => [:spec]"
      end
    end

    private

    def override_path_for_tests
      if ENV['TESTING']
        support_bin = File.expand_path(File.join('..', '..', '..', 'features', 'support', 'bin'))
        "PATH=#{support_bin}:$PATH"
      end
    end

    def factories_spec_rake_task
      IO.read find_in_source_paths('factories_spec_rake_task.rb')
    end

    def simplecov_init
      IO.read find_in_source_paths('simplecov_init.rb')
    end
  end
end
